pl_bolts/datamodules/emnist_datamodule.py

```diff
@@ -0,0 +1,178 @@
+from typing import Any, Callable, Optional, Union
+
+from pl_bolts.datamodules.vision_datamodule import VisionDataModule
+from pl_bolts.datasets import EMNIST
+from pl_bolts.utils import _TORCHVISION_AVAILABLE
+from pl_bolts.utils.warnings import warn_missing_pkg
+
+if _TORCHVISION_AVAILABLE:
+    from torchvision import transforms as transform_lib
+else:  # pragma: no cover
+    warn_missing_pkg('torchvision')
+
+
+class EMNISTDataModule(VisionDataModule):
+    """
+    .. figure:: https://user-images.githubusercontent.com/4632336/123210742-4d6b3380-d477-11eb-80da-3e9a74a18a07.png
+        :width: 400
+        :alt: EMNIST
+    Specs:
+        - 6 splits: ``byclass``, ``bymerge``,
+                    ``balanced``, ``letters``,
+                    ``digits`` and ``mnist``.
+
+        - Table:
+
+          | Split Name   | Classes | No. Training | No. Testing | Validation | Total   |
+          |--------------|---------|--------------|-------------|------------|---------|
+          | ``byclass``  | 62      | 697,932      | 116,323     | No         | 814,255 |
+          | ``bymerge``  | 47      | 697,932      | 116,323     | No         | 814,255 |
+          | ``balanced`` | 47      | 112,800      | 18,800      | Yes        | 131,600 |
+          | ``digits``   | 10      | 240,000      | 40,000      | Yes        | 280,000 |
+          | ``letters``  | 37      | 88,800       | 14,800      | Yes        | 103,600 |
+          | ``mnist``    | 10      | 60,000       | 10,000      | Yes        | 70,000  |
+
+          source: https://arxiv.org/pdf/1702.05373.pdf [Table-II]
+
+        - For each split:
+          - Each image is (1 x 28 x 28)
+    Standard EMNIST, train, val, test-splits and transforms
+    Transforms::
+        emnist_transforms = transform_lib.Compose([
+            transform_lib.ToTensor()
+        ])
+    Example::
+        from pl_bolts.datamodules import EMNISTDataModule
+        dm = EMNISTDataModule('.')
+        model = LitModel()
+        Trainer().fit(model, datamodule=dm)
+    """
+    name = "emnist"
+    dataset_cls = EMNIST
+    dims = (1, 28, 28)
+
+    def __init__(
+        self,
+        data_dir: Optional[str] = None,
+        split: str = 'mnist',
+        val_split: Union[int, float] = 0.2,
```

Comment:
This is only a suggestion, but how about setting `val_split` to `None` by default, and in `__init__(...)`, assigning a `int` or `float` value defined in the paper to `val_split`?

```suggestion
        val_split: Optional[int, float] = None,
```

Reply:
@akihironitta `MNISTDataModule` and `BinaryMNISTDataModule` uses a `val_split = 0.2`. So, I suggest we keep the same as a default argument for `EMNISTDataModule` and `BinaryEMNISTDataModule`. This will allows for less stringent requirement for the user to visit documentation while using the API: *This will at least create train-val-test datasets, each of finite length.* 

> *Think of all the packages that follow sklearn's api format; a typical user, who has used sklearn's API, would know almost immediately the design of another similar API, even before using it*. 

However, I like your suggestion of using `None` as a value to flag the use of **default** train-val-test splitting of the EMNIST dataset. In fact, the current implementation, uses `val_split is None` as the conditional to fetch the validation dataset size (this requires the EMNIST_METADATA).

Additionally, I have added another layer of abstraction in `EMNISTDataModule` (I have not yet implemented the same for `BinaryEMNISTDataModule` though): 

```python
# _DEFAULT_NO_VALIDATION_VAL_SPLIT: This is the `val_split` to use when
# "validation = False" for a given split in the metadata and the user-input
# for `val_split` is `None`.
_DEFAULT_NO_VALIDATION_VAL_SPLIT: Union[int, float] = 0.
```

## About using `Optional` for `val_split`

If you want to use `Optional` for type-hinting `val_split`, I think you have a typo in what you suggested.  

```Diff
-        val_split: Optional[int, float] = None,
+        val_split: Optional[Union[int, float]] = 0.2,
```

The `Optional[...]` takes only a single type as argument, and so, if I understood you correctly, we could use either of the following (see [this stackoverflow discussion][#so-question] for more details):

[#so-question]: https://stackoverflow.com/questions/51710037/how-should-i-use-the-optional-type-hint

```python
# method-1: Optional + Union
val_split: Optional[Union[int, float]] = 0.2 # or, None

# method-2: Union + None
val_split: Union[int, float, None] = 0.2 # or, None
```

:star: :question: Which method you think is better: `method-1` or `method-2`?

Reply:
Sure. Let me check my understanding of the behaviour you're suggesting here. Are the following statements correct?:

- `val_split` is set to `0.2` by default to keep consistency with sklearn.
- Passing `None` to `val_split` uses the train-val split defined in the paper for the balanced splits.
- Passing `None` to `val_split` uses `0` (no validation set) for non-balanced splits.

---

As for the typing `Optional`, you're right. It's just a typo :sweat_smile: